### PR TITLE
bigint2 acceleration update

### PR DESCRIPTION
**Pull request by austinabell**
Perhaps we want to start a new branch, to avoid pre 1.2 versions from being borked, just pointing this PR to somewhere for now.

### CODE REVIEW

**Review by tzerrell — APPROVED** [review]
Looks good!

**Comment by austinabell** [comment]
@nategraf do you have any preferences on where the bigint2 accelerator patches live as far as the branch name or how the release is tagged? Assume there should be a new branch, to ensure no breakages from using the `risczero` branch through cargo for those not on 1.2 risc0. That may not be an issue, though, as we always recommend using the release tag.

cc @pdg744

**Comment by nategraf** [comment]
Running `cargo risczero test` here, I hit can issue with the `proc-macro2` crate, and bumped it's version to work with rust 1.81. Then I got tests to run and mostly pass, but hit the following failure:

```
Error: Can't divide by zero
test ecdsa::tests::wycheproof::wycheproof ... %
```

Unfortunately not super easy to debug, since we don't get a backtrace of anything. For context `wycheproof` is a test suite for crytography including ECDSA that includes a lot of edge cases. https://github.com/C2SP/wycheproof

File: k256/src/arithmetic/mul.rs [hunk 1]
```diff
@@ -339,6 +339,40 @@ impl LinearCombination<[(ProjectivePoint, Scalar)]> for ProjectivePoint {
...
+
+    let mut result = ec::AffinePoint::new_unchecked([0u32; 8], [0u32; 8]);
+    affine.mul(&scalar, &mut result);
+    let mut result_buffer = ec::AffinePoint::new_unchecked([0u32; 8], [0u32; 8]);
+    let mut mul_buffer = ec::AffinePoint::new_unchecked([0u32; 8], [0u32; 8]);
+    for (point, scalar) in xks_iter {
+        point.mul(&scalar, &mut mul_buffer);
+        // TODO experiment with alternating buffers for perf
+        result.add(&mul_buffer, &mut result_buffer);
+        core::mem::swap(&mut result, &mut result_buffer);
```
Comment (nategraf) [inline k256/src/arithmetic/mul.rs]: Does this swap the actual values, or just the pointers? It might be worth defining `result_buffer` as type `&mut ec::AffinePoint` and then swapping the references here. Not 100% sure that will work, will produce better results, and that Rust will not complain.

**Review by nategraf — COMMENTED** [review]
Looks good to me. Just some questions to resolve around releasing (e.g. changing the bigint2 reference to a released version)

**Comment by austinabell** [comment]
> Just some questions to resolve around releasing (e.g. changing the bigint2 reference to a released version)

Yeah, will be in draft until risc0 bigint2 release, and I will update

**Review by tzerrell — APPROVED** [review]

File: k256/src/arithmetic/mul.rs (hunk 1 above)
Comment (austinabell) [reply k256/src/arithmetic/mul.rs]: That's not possible, because you need to access the value immutable in-between where the buffers are defined and where this swap is.

I understand what you're saying, and will think more about this and resolve this TODO (or measure with my initial plan) before it gets merged.

Comment (austinabell) [reply k256/src/arithmetic/mul.rs]: I was wrong, it is possible to just use `mem::swap` on the pointers, and it gives roughly the same perf as just alternating buffers, but much cleaner. Assume it's roughly the same as alternating because swapping the address in memory is likely as cheap as %2 checks (just a guess)

[`a12c63a` (#6)](https://github.com/risc0/RustCrypto-elliptic-curves/pull/6/commits/a12c63a066344d2990c93e5cc4e915019b90b6ab)

**Comment by austinabell** [comment]
tested with `cargo risczero test` and tested patching on a crate with latest changes -- should be good to go from my end

File: k256/src/arithmetic/projective.rs [hunk 2]
```diff
@@ -108,6 +144,7 @@ impl ProjectivePoint {
         let yz_pairs = ((self.y + &self.z) * &(other.y + &other.z)) + &n_yy_zz;
         let xz_pairs = ((self.x + &self.z) * &(other.x + &other.z)) + &n_xx_zz;
 
+        // TODO can remove these to simplify patch (from bigint1)
```
Comment (austinabell) [inline k256/src/arithmetic/projective.rs]: Can remove this comment to simplify applying just bigint1 patch and this new one (maybe less conflict)

Comment (austinabell) [reply k256/src/arithmetic/projective.rs]:
```suggestion
```

Comment (austinabell) [reply k256/src/arithmetic/projective.rs]: Made the change, but maybe an open question of what we do with that logic now given it's obsolete with bigint2

File: k256/src/arithmetic/projective.rs [hunk 3]
```diff
@@ -220,6 +257,7 @@ impl ProjectivePoint {
         ret
     }
 
+    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
     /// Doubles this point.
     #[inline]
```
Comment (flaub) [inline k256/src/arithmetic/projective.rs]:
```suggestion
    /// Doubles this point.
    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
    #[inline]
```

Comment (flaub) [reply k256/src/arithmetic/projective.rs]: I only marked one of this pattern, but I saw a few more

File: k256/src/arithmetic/projective.rs [hunk 4]
```diff
@@ -92,6 +100,34 @@ impl ProjectivePoint {
...
+    fn add_mixed(&self, other: &AffinePoint) -> ProjectivePoint {
+        let a = projective_to_affine(self);
+        let b = affine_to_bigint2_affine(other);
+        let mut result = ec::AffinePoint::IDENTITY;
+        a.add(&b, &mut result);
+        affine_to_projective(&result)
+    }
+
+    #[cfg(all(target_os = "zkvm", target_arch = "riscv32"))]
+    /// Doubles this point.
```
Comment (austinabell) [inline k256/src/arithmetic/projective.rs]:
```suggestion
    /// Doubles this point.
    #[cfg(all(target_os = "zkvm", target_arch = "riscv32"))]
```

File: k256/src/arithmetic/projective.rs [hunk 5]
```diff
@@ -160,6 +196,7 @@ impl ProjectivePoint {
         }
     }
 
+    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
     /// Returns `self + other`.
```
Comment (austinabell) [inline k256/src/arithmetic/projective.rs]:
```suggestion
    /// Returns `self + other`.
    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
```

File: k256/src/arithmetic/projective.rs [hunk 6]
```diff
@@ -92,6 +100,34 @@ impl ProjectivePoint {
...
+    #[inline]
+    pub fn double(&self) -> ProjectivePoint {
+        let a = projective_to_affine(self);
+        let mut result = ec::AffinePoint::IDENTITY;
+        a.double(&mut result);
+        affine_to_projective(&result)
+    }
+
+    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
     /// Returns `self + other`.
```
Comment (austinabell) [inline k256/src/arithmetic/projective.rs]:
```suggestion
    /// Returns `self + other`.
    #[cfg(not(all(target_os = "zkvm", target_arch = "riscv32")))]
```

File: k256/src/arithmetic/projective.rs (hunk 3 above)
Comment (austinabell) [reply k256/src/arithmetic/projective.rs]: got them in commits below :)

File: k256/src/arithmetic/projective.rs (hunk 2 above)
Comment (tzerrell) [reply k256/src/arithmetic/projective.rs]: In my opinion we should just drop it. It's unreachable code, it's not useful with the current architecture, and if we discover we _do_ actually need it later for some reason, git will remember the past for us.

Comment (austinabell) [reply k256/src/arithmetic/projective.rs]: Fair enough, was assuming it might make applying just this commit more difficult in the future, but probably not a case where it matters

**Comment by austinabell** [comment]
Created release branches for all the versions we have released previously, changed base to 13.3.

Once this is approved, I can either squash these commits manually and apply to other branches, or once this is approved then squash merged, I can apply that commit to other versions.

**Review by tzerrell — APPROVED** [review]
